packages/aws-cdk-lib/aws-eks/lib/managed-nodegroup.ts

```diff
@@ -479,6 +508,21 @@ function isGpuInstanceType(instanceType: InstanceType): boolean {
   return ['p', 'g', 'inf'].includes(family);
 }
 
+/**
+ * This function check if the instanceType is supported by Windows AMI.
+ * https://docs.aws.amazon.com/eks/latest/userguide/windows-support.html
+ * @param instanceType The EC2 instance type
+ */
+function isWindowsSupportedInstanceType(instanceType: InstanceType): boolean {
+  // capture the family, generation, capabilities, and size portions of the instance type id
+  const instanceTypeFormat = instanceType.toString().match(/^[a-z]+\d{1,2}[a-z]*\.[a-z0-9]+$/);
+  if (instanceTypeFormat == null) {
+    throw new Error('Malformed instance type identifier');
+  }
+  const forbiddenInstanceTypes:RegExp = /^(?:c3|c4|d2|i2|m4|m6a|r3)\.[a-z0-9]+$/;
```

Comment:
This kind of thing breaks the encapsulation that `InstanceType` is supposed to provide. I think it's better to move this logic inside `InstanceType` somehow. An idea off the cuff:

```ts
public sameInstanceClassAs(other: InstanceType): boolean {
  // parse this.instanceTypeIdentifier and other.instanceTypeIdentifier 
  // and compare the parts before the '.'
}
```

but you may come up with a more elegant solution along these lines.

Reply:
This function was modeled after `function isGpuInstanceType(instanceType: InstanceType): boolean` ...  

Expanding the capability of InstanceTypes with a function like 

```TypeScript
    public sameInstanceClassAs(other: InstanceType): boolean {
        const instanceClass: RegExp = /^([a-z]+\d{1,2}[a-z]*)\.([a-z0-9]+)$/;
        const instanceClassId = this.instanceTypeIdentifier.match(instanceClass);
        const otherInstanceClassId = other.toString().match(instanceClass);
        if (instanceClassId == null || otherInstanceClassId == null) {
          throw new Error('Malformed instance type identifier');
        }
        return instanceClassId[1] === otherInstanceClassId[1];
    }
```

and co-opting the new method is a good idea. It does not help solve the immediate problem of the lack of support for Windows Containers in CDK for EKS. Perhaps we can merge this change and then create another issue and PR to clean up the instance type checks?

Reply:
The purpose of this pull request is to address issue https://github.com/aws/aws-cdk/issues/24803 which deals strictly with adding values to NodegroupAmiType that are missing in CDK but supported in CloudFormation. I set up some guard rails to ensure the users of CDK would not inadvertently select an unsupported instance type for their Windows NodeGroups. The guard rails function present in the PR mimics the preexisting guard rails function for GPU-based NodeGroups. The proposed patch shouldn't be faulted for re-using a pattern already present in the codebase to solve an immediate problem.

I am not denying that those functions could be improved. It is a good thing you suggest.

However, I propose that this PR be merged as-is, and a separate issue for cleaning up the instance type checking routines be submitted.

Each PR would represent an atomic concern. This is a good practice.